cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/TransformRunner.java

```diff
@@ -129,7 +129,8 @@ public TransformRunner(MapReduceTaskContext<Object, Object> context,
     Set<StageInfo> reducers = pipelinePhase.getStagesOfType(BatchAggregator.PLUGIN_TYPE, BatchJoiner.PLUGIN_TYPE);
     if (!reducers.isEmpty()) {
       String reducerName = reducers.iterator().next().getName();
-      if (pipelinePhase.getSinks().contains(reducerName)) {
+      JobContext hadoopContext = context.getHadoopContext();
+      if (pipelinePhase.getSinks().contains(reducerName) || hadoopContext instanceof Mapper.Context) {
```

Comment:
ok I read through the code again, this is not where the fix should be and will in fact introduce bugs. Doesn't this cause unit tests to fail?

Instead, you need to change the TransformRunner to call this method after it does the phase subsets.